### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by AdaRoseCannon**
Can, @railsagainstignorance @JakeChampion and @seanmtracey, take a look. 

The tests can take a little long to run because due to the nature of scheduling it requires for us to wait for the minute hand to change for one of the tests. 

They also raised a few of small issues which I have fixed:
1. if you ran viewer.removeAllListeners('change') it would stop syncing up.
2. if you made an update with no item array e.g. `{id:12345}` rather than `{id:12345, items: []}` it would error.
3. if you ran it on a browser without native `Array.find` it would error.

To run the tests:

```
npm install
npm run test
```

The tests work by:
1. Compiling the libary to ES5 (so the distribution version is up to date)
2. Builds a standalone version of the ES5 library using browserify
3. An express http server is started to serve the tests and provide a sockets.io echo server.
4. Phantomjs is started and loads up the test html page from the http server.
5. Mocha and Chai are pulled down from the cdn
6. The tests then use the socket echo to spoof api responses and test for the correct response.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 'use strict';
 /* global io */
 
+require("babel-polyfill");
```
Comment (JakeChampion) [inline lib/index.js]: Why is this required? It is a huge polyfill :/

File: test/test.html [hunk 2]
```diff
@@ -0,0 +1,23 @@
+<html>
+	<head>
+		<meta charset="utf-8">
+		<!-- encoding must be set for mocha's special characters to render properly -->
```
Comment (JakeChampion) [inline test/test.html]: I don't think we need the comment, encoding should always be declared first thing in the `head` tag to avoid some forms of XSS attacks.

File: test/tests/viewer.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
+'use strict';
+/* global Viewer, describe, it, chai, afterEach */
+var expect = chai.expect;
```
Comment (JakeChampion) [inline test/tests/viewer.js]: Does this project make use of the latest version of OBT? `var` is deprecated in favour of `const` & `let`

File: test/tests/viewer.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+	// remove all listeners
+	afterEach(function() {
+		viewer.removeAllListeners('change');
+		viewer.removeAllListeners('update');
+		viewer.removeAllListeners('reload');
+	});
+
+	it('Should reload when recieves reload', function (done) {
+		viewer.once('reload', function() { done() });
```
Comment (JakeChampion) [inline test/tests/viewer.js:82]: Instead of wrapping `done` in an anonymous function, you could pass `done` directly as a value to `once`:

``` js
viewer.once('reload', done);
```

File: test/tests/viewer.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+	it('Should not update when only id updated', function (done) {
+		function err() { throw Error('Should not change, no items added.') }
+		viewer.once('change', err);
+		mockAPI('updateId');
+		viewer.socket.once('update', function () {
+			setTimeout(function() {
+				viewer.removeListener('change', err);
+				done();
+			}, 200);
```
Comment (JakeChampion) [inline test/tests/viewer.js:94]: Does this need to be in a `setTimeout`?

File: test/tests/viewer.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+				expect(viewer.getUrl()).to.be.undefined;
+				done();
+			});
+		});
+	});
+
+	it('Should show scheduled pages at the correct time.', function (done) {
+
+		// this will take at most one minute as it succeeds when the clock ticks over
+		this.timeout(61000);
```
Comment (JakeChampion) [inline test/tests/viewer.js:143]: Is there anything like [timecop](https://github.com/travisjeffery/timecop) for JS?

File: test/tests/viewer.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+				done();
+			} else {
+				throw Error('Updated too soon.');
+			}
+		});
+		mockAPI('updateUrl');
+	});
+
+	it('Should not refresh the page if the url did not change', function (done) {
+		delete mockAPIs.updateUrl.items[0].dateTimeSchedule;
```
Comment (JakeChampion) [inline test/tests/viewer.js]: As all the tests are using the same mock object, if this test ran first it would make others fails (the `dateTimeSchedule` would be undefined). I would recommend cloning the mockAPI in each test. It is safer to repeat code in tests than to DRY things up.

File: test/tests/viewer.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+		viewer.once('change', function () {
+			throw Error('change was fired even though url did not change.');
+		});
+		viewer.socket.on('update', function () {
+
+			// Make sure no change has happened after an update has been recieved.
+			setTimeout(done, 200);
```
Comment (JakeChampion) [inline test/tests/viewer.js]: I'm not sure this will ever get called as there is a `done()` on line 132.

**Comment by JakeChampion** [comment]
Finished my review. Added comments.

File: lib/index.js (hunk 1 above)
Comment (AdaRoseCannon) [inline lib/index.js]: :+1: For Array.find, i'll use indexOf instead.

File: test/test.html (hunk 2 above)
Comment (AdaRoseCannon) [inline test/test.html]: Sorry this was from the mocha phantomjs boilerplate

File: test/tests/viewer.js (hunk 3 above)
Comment (AdaRoseEdwards) [inline test/tests/viewer.js]: This needs to run in phantomjs which is essentially an older version of chrome. It won't run with any es6.

File: test/tests/viewer.js (hunk 4 above)
Comment (AdaRoseCannon) [inline test/tests/viewer.js:82]: That also binds done to viewer. Don't want to break any mocha internals.

File: test/tests/viewer.js (hunk 5 above)
Comment (AdaRoseCannon) [inline test/tests/viewer.js:94]: Yes so that there is time for the api request to be spoofed.

File: test/tests/viewer.js (hunk 6 above)
Comment (AdaRoseCannon) [inline test/tests/viewer.js:143]: Not that I am aware of. Also time is hard and I would rather test actual behaviour if possible.

File: test/tests/viewer.js (hunk 7 above)
Comment (AdaRoseCannon) [inline test/tests/viewer.js]: Mocha runs each 'it' in order. This test itself relies on the test before it. They are in a single `describe` it seems appropriate that they can rely on order.

Although I am in agreement that perhaps putting copying each individual api request in the request map at the top.

**Making it so that each test does not rely on any of the others.**

File: test/tests/viewer.js (hunk 8 above)
Comment (AdaRoseCannon) [inline test/tests/viewer.js]: Good catch. :+1:

**Comment by seanmtracey** [comment]
Looks good to me. Tests pass too :)
